src/mpi_graph_io.cpp

```diff
@@ -0,0 +1,143 @@
+#include <gdsb/mpi_graph_io.h>
+
+#include <cstddef>     // for: offsetof()
+#include <type_traits> // for: is
+
+namespace gdsb
+{
+
+namespace mpi
+{
+
+FileWrapper::FileWrapper(std::filesystem::path const& file_path, int const mode)
+{
+    int error = MPI_File_open(MPI_COMM_WORLD, file_path.c_str(), mode, MPI_INFO_NULL, &m_file);
+
+    if (error != MPI_SUCCESS)
+    {
+        throw std::runtime_error("Could not open file using MPI routines.");
+    }
+}
+
+FileWrapper::FileWrapper(std::filesystem::path const& file_path, bool const overwrite, int const mpi_root_process, int const mode)
+{
+    int const test_open_error = MPI_File_open(MPI_COMM_WORLD, file_path.c_str(), mode | MPI_MODE_EXCL, MPI_INFO_NULL, &m_file);
+    if (test_open_error != MPI_SUCCESS)
+    {
+        int rank;
+        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+        if (rank == mpi_root_process)
+        {
+            MPI_File_delete(file_path.c_str(), MPI_INFO_NULL);
```

Comment:
this code seems confusing without additional comments / explanation

Reply:
I added some comments and improved the tests for the exclusive open option.